api/nodes/serializers.py

```diff
@@ -127,6 +128,133 @@ def update(self, instance, validated_data):
+        current_user = self.context['request'].user
+        auth = Auth(current_user)
+        node = self.context['view'].get_node()
+        user = User.load(validated_data['_id'])
+        is_admin_current = node.has_permission(current_user, 'admin')
+        if not user:
+            raise NotFound('User with id {} cannot be found.'.format(validated_data['_id']))
+        elif user in node.contributors:
+            raise ValidationError('User {} already is a contributor.'.format(user.username))
+        bibliographic = validated_data['bibliographic'] == "True"
+        permission_field = validated_data['permission']
+        node.add_contributor(contributor=user, auth=auth, save=True)
+        self.set_visibility(bibliographic, user, node, is_admin_current)
+        if permission_field != '':
+            self.set_permissions(permission_field, user, node, is_admin_current)
+        user.node_id = node._id
+        user.bibliographic = node.get_visible(user)
+        return user
+
+    def set_visibility(self, bibliographic, user, node, is_admin_current, is_current=None):
+        if is_admin_current and bibliographic:
+            node.set_visible(user, True, save=True)
+        elif not bibliographic:
+            if len(node.visible_contributors) == 1 and is_current:
+                raise ValidationError('Must have at least one visible contributor')
+            elif is_admin_current or is_current:
+                node.set_visible(user, False, save=True)
+        else:
+            raise PermissionDenied()
+
+    def set_permissions(self, field, user, node, is_admin, edit=None):
+        if is_admin or edit:
+            if field == 'admin':
+                    node.set_permissions(user, ['read', 'write', 'admin'])
+            if self.has_multiple_admins(node) or not edit:
+                if field == 'write':
+                    node.set_permissions(user, ['read', 'write'])
+                elif field == 'read':
+                    node.set_permissions(user, ['read'])
+            else:
+                raise ValidationError('Must have at least one admin contributor')
+        elif field == 'read' and node.has_permission(user, 'write'):
+            node.set_permissions(user, ['read'])
+        else:
+            raise PermissionDenied()
+        node.save()
+
+    def has_multiple_admins(self, node):
+        has_one_admin = False
+        for contributor in node.contributors:
+            if node.has_permission(contributor, 'admin'):
+                if has_one_admin:
+                    return True
+                has_one_admin = True
+        return False
+
+
+class NodeContributorDetailSerializer(NodeContributorsSerializer):
+
+    id = ser.CharField(read_only=True, source='_id')
+
+    # Overridden to allow blank for user to not change status
+    permission = ser.ChoiceField(choices=['admin', 'read', 'write'], write_only=True, allow_blank=True)
+
+    # Overridden to remove contributor detail link
```

Comment:
I think it's pretty standard for detail items to have links to themselves. Verify by checking out the node list and node detail. There should be a self link in both of those. If so, go ahead and take out this override.